### PR TITLE
feat: add json serialisation support for golang 

### LINKED ISSUE #1842: [BUG] go generator is unable to handle unions types correctly.
### Describe the bug.

Go generator cannot generate models for types that could be two separate types simultaneously, for example, a servers object that could be a list of server objects or a reference object. 

### Expected behavior

Go generator can generate models that support union types or replicate the functionality of union types in golang, with respect to what is possible in golang.

### Screenshots

Screenshot 2024-02-28 at 2 39 22 PM

`servers` could be a reference type as well. 

### How to Reproduce

use `modelina` to generate models for the [AsyncAPI JSON scheme](https://github.com/asyncapi/spec-json-schemas/blob/master/schemas/3.0.0-without-%24id.json)

### 🥦 Browser

Mozilla Firefox

### 👀 Have you checked for similar open issues?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/asyncapi/.github/blob/master/CONTRIBUTING.md)

### Are you willing to work on this issue ?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Souvikns**
## Description
This PR is adding JSON serialization support by adding struct tags while generating go structs, by creating a new `GO_PRESET`.

```go
package models

type AsyncApi_3Dot_0Dot_0SchemaDot struct {
  Asyncapi string `json:"asyncapi"`
  Id string `json:"id"`
  Info *Info `json:"info"`
  Servers map[string]interface{} `json:"servers"`
  DefaultContentType string `json:"defaultContentType"`
  Channels map[string]interface{} `json:"channels"`
  Operations map[string]interface{} `json:"operations"`
  Components *Components `json:"components"`
  AdditionalProperties map[string]interface{} `json:"-"`
}
```

Also adds custom marshaler functions for enums when generating `struct-tags`

```go
func (op *ApiKeyIn) UnmarshalJSON(raw []byte) error {
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return err
	}
	*op = ValuesToApiKeyIn[v]
	return nil
}

func (op ApiKeyIn) MarshalJSON() ([]byte, error) {
	return json.Marshal(op.Value())
} 
          

```

## Related Issue

Fixes https://github.com/asyncapi/modelina/issues/299

## Checklist
- [x] The code follows the project's coding standards and is properly linted (`npm run lint`).
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated to reflect the changes.
- [x] All tests pass successfully locally.(`npm run test`).

### CODE REVIEW

File: test/generators/go/presets/CommonPreset.spec.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  GoGenerator,
+  GO_COMMON_PRESET,
+  GoCommonPresetOptions
+} from '../../../../src/generators';
+
+describe('Go_COMMON_PRESET', () => {
+  let generator: GoGenerator;
+  beforeEach(() => {
+    generator = new GoGenerator({
+      presets: [{ preset: GO_COMMON_PRESET, options: { addJsonTag: true } }]
```
Comment (Souvikns) [inline test/generators/go/presets/CommonPreset.spec.ts:11]: @jonaslagoni I am having some issues here with `GO_COMMON_PRESET` 
Screenshot 2024-03-01 at 7 18 42 PM

Comment (jonaslagoni) [reply test/generators/go/presets/CommonPreset.spec.ts:11]: Go into /src/generators/go/GoPreset.ts and change the following lines:
```
export type GoPreset<O = any> = Preset<{
  struct: StructPresetType<O>;
  enum: EnumPresetType<O>;
}>;

export const GO_DEFAULT_PRESET: GoPreset<GoOptions> = {
  struct: GO_DEFAULT_STRUCT_PRESET,
  enum: GO_DEFAULT_ENUM_PRESET
};
```

Its a bug with the preset types 🙂

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Only one comment for the implementation 👍 Looking good 💪 

Remember to do parts 4 and 5 of https://github.com/asyncapi/modelina/blob/master/docs/contributing.md#adding-a-new-preset

File: examples/go-json-tags/package.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "config" : { "example_name" : "TEMPLATE" },
```
Comment (jonaslagoni) [inline examples/go-json-tags/package.json:2]: This needs to change

**Review by jonaslagoni — CHANGES_REQUESTED** [review]

**Review by jonaslagoni — APPROVED** [review]
LGTM 👍

**Pull request merged**